lib/ClangImporter/SwiftDeclSynthesizer.cpp

```diff
@@ -2534,128 +2534,167 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
-      clangCtx, const_cast<clang::CXXRecordDecl *>(cxxRecordDecl),
-      cxxRecordDecl->getLocation(),
-      clang::DeclarationNameInfo(funcNameToSynthesize,
-                                 cxxRecordDecl->getLocation()),
-      funcTypeToSynthesize,
-      clangCtx.getTrivialTypeSourceInfo(funcTypeToSynthesize), clang::SC_Static,
-      /*UsesFPIntrin=*/false, /*isInline=*/true,
-      clang::ConstexprSpecKind::Unspecified, cxxRecordDecl->getLocation());
-  assert(synthesizedCxxMethodDecl &&
-         "Unable to synthesize static factory for c++ foreign reference type");
-  synthesizedCxxMethodDecl->setAccess(clang::AccessSpecifier::AS_public);
-
-  if (!hasImmortalAttrs(cxxRecordDecl)) {
-    clang::SwiftAttrAttr *returnsRetainedAttrForSynthesizedCxxMethodDecl =
-        clang::SwiftAttrAttr::Create(clangCtx, "returns_retained");
+
+  for (clang::CXXConstructorDecl *selectedCtorDecl : cxxRecordDecl->ctors()) {
+    if (selectedCtorDecl->isDeleted() ||
+        selectedCtorDecl->getAccess() == clang::AS_private ||
+        selectedCtorDecl->getAccess() == clang::AS_protected ||
+        selectedCtorDecl->isCopyOrMoveConstructor())
+      continue;
+
+    unsigned int ctorParamCount = selectedCtorDecl->getNumParams();
+    std::string funcName = "__returns_" + cxxRecordDecl->getNameAsString();
+    if (ctorParamCount > 0)
+      funcName += "_" + std::to_string(ctorParamCount) + "_params";
+    clang::IdentifierInfo *funcNameToSynthesize = &clangIdents.get(funcName);
+
+    llvm::SmallVector<clang::QualType, 4> paramTypes;
+    for (const auto *param : selectedCtorDecl->parameters())
+      paramTypes.push_back(param->getType());
+    clang::FunctionProtoType::ExtProtoInfo EPI;
+    clang::QualType funcTypeToSynthesize =
+        clangCtx.getFunctionType(cxxRecordPtrTy, paramTypes, EPI);
+
+    clang::CXXMethodDecl *synthesizedCxxMethodDecl =
+        clang::CXXMethodDecl::Create(
+            clangCtx, const_cast<clang::CXXRecordDecl *>(cxxRecordDecl),
+            cxxRecordDeclLoc,
+            clang::DeclarationNameInfo(funcNameToSynthesize, cxxRecordDeclLoc),
+            funcTypeToSynthesize,
+            clangCtx.getTrivialTypeSourceInfo(funcTypeToSynthesize),
+            clang::SC_Static, /*UsesFPIntrin=*/false, /*isInline=*/true,
+            clang::ConstexprSpecKind::Unspecified, cxxRecordDeclLoc);
+    assert(
+        synthesizedCxxMethodDecl &&
+        "Unable to synthesize static factory for c++ foreign reference type");
+    synthesizedCxxMethodDecl->setAccess(clang::AccessSpecifier::AS_public);
+
+    llvm::SmallVector<clang::ParmVarDecl *, 4> synthesizedParams;
+    for (unsigned int i = 0; i < ctorParamCount; ++i) {
+      auto *origParam = selectedCtorDecl->getParamDecl(i);
+      auto *param = clang::ParmVarDecl::Create(
+          clangCtx, synthesizedCxxMethodDecl, cxxRecordDeclLoc,
+          cxxRecordDeclLoc, origParam->getIdentifier(), origParam->getType(),
```

Comment:
`origParam->getIdentifier()` can be a bit problematic. Consider the following case:
```
struct X {
  X(T);
};
```

Here, we specified the type for the argument but did not give it a name. In this case there is no identifier and we would not be able to refer to this parameter. Let's add a test case to test this scenario. Also add some code that generates a synthesized name in case the original parameter has no identifier.